src/abstractdataframe/abstractdataframe.jl

```diff
@@ -685,78 +685,66 @@ Base.vcat(df::AbstractDataFrame) = df
 Base.vcat(dfs::AbstractDataFrame...) = vcat(AbstractDataFrame[dfs...])
 
 Base.vcat(dfs::Vector{Void}) = dfs
+
 function Base.vcat{T<:AbstractDataFrame}(dfs::Vector{T})
     isempty(dfs) && return DataFrame()
-    coltyps, colnams, similars = _colinfo(dfs)
-
     res = DataFrame()
-    Nrow = sum(nrow, dfs)
-    for j in 1:length(colnams)
-        colnam = colnams[j]
-        col = similar(similars[j], coltyps[j], Nrow)
-
-        i = 1
-        for df in dfs
-            if haskey(df, colnam)
-                copy!(col, i, df[colnam])
+    nrows = sum(nrow, dfs)
+    for colnam in unique(Base.flatten(names.(dfs)))
+        k = Bool[haskey(df, colnam) for df in dfs]
+        C = Base.return_types(vcat, [typeof(dfs[i][colnam]) for i in 1:length(dfs) if k[i]])
+
+        if length(C)==1 && isleaftype(C[1])
+            if _isnullable(C[1])
+              NC = C[1]
+            else
+              NC = NullableArray{eltype(C[1])}
```

Comment:
I used the value from `return_types` instead of `NullableArray{promote_type(eltypes(c)}` as @nalimilan suggested.